File: mrscraper_mcp/status.py
"""CLI-compatible account-status and date helpers."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
import re
from typing import Any

_DURATION_PATTERN = re.compile(r"^(\d+)(m|h|d|w)$")


def format_api_date(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")


def parse_status_date(
    value: str | None,
    now: datetime | None = None,
    fallback_duration: str = "24h",
) -> datetime:
    reference = now or datetime.now(timezone.utc)
    if reference.tzinfo is None:
        reference = reference.replace(tzinfo=timezone.utc)

    input_value = (value or fallback_duration).strip().lower()
    if input_value == "now":
        return reference

    match = _DURATION_PATTERN.fullmatch(input_value)
    if match:
        amount = int(match.group(1))
        unit = match.group(2)
        seconds = {
            "m": 60,
            "h": 3_600,
            "d": 86_400,
            "w": 604_800,
        }[unit]
        return reference - timedelta(seconds=amount * seconds)

    try:
        normalized = (
            input_value[:-1] + "+00:00" if input_value.endswith("z") else input_value
        )
        parsed = datetime.fromisoformat(normalized)
    except ValueError as exc:
        raise ValueError(
            f'Invalid date "{value}". Use ISO 8601, "now", or a duration such as 24h or 7d.'
        ) from exc
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _number_or_zero(value: Any) -> float | int:
    if isinstance(value, bool):
        return int(value)
    try:
        number = float(value)
    except (TypeError, ValueError):
        return 0
    return int(number) if number.is_integer() else number


def summarize_subscription_account(account: dict[str, Any]) -> dict[str, Any]:
    token_limit = _number_or_zero(account.get("tokenLimit"))
    token_usage = _number_or_zero(account.get("tokenUsage"))
    user = account.get("user") if isinstance(account.get("user"), dict) else {}
    remaining = max(0, token_limit - token_usage)
    usage_percent = round((token_usage / token_limit) * 100, 2) if token_limit else 0

    return {
        "subscription_status": account.get("stripeStatus"),
        "enterprise": bool(account.get("isEnterprise")),
        "token_usage": token_usage,
        "token_limit": token_limit,
        "token_remaining": remaining,
        "usage_percent": usage_percent,
        "rate_limit": _number_or_zero(account.get("rateLimit")),
        "rate_ttl": _number_or_zero(account.get("rateTtl")),
        "auto_renew": bool(account.get("isAutoRenew")),
        "ends_at": account.get("endsAt"),
        "user": {
            "name": user.get("name"),
            "email": user.get("email"),
            "verified": bool(user.get("isVerified")),
        },
    }

File: src/status.ts
function numberOrZero(value: unknown): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : 0;
}

function recordOrEmpty(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

export function formatApiDate(date: Date): string {
  if (!(date instanceof Date) || Number.isNaN(date.getTime())) {
    throw new Error("Invalid date");
  }
  return date
    .toISOString()
    .replace("T", " ")
    .replace(/\.\d{3}Z$/, "");
}

export function parseStatusDate(
  value: string | undefined,
  now = new Date(),
  fallbackDuration = "24h",
): Date {
  const input = (value || fallbackDuration).trim().toLowerCase();
  if (input === "now") return new Date(now);

  const relative = /^(\d+)(m|h|d|w)$/.exec(input);
  if (relative) {
    const amount = Number(relative[1]);
    const units: Record<string, number> = {
      m: 60_000,
      h: 3_600_000,
      d: 86_400_000,
      w: 604_800_000,
    };
    return new Date(now.getTime() - amount * (units[relative[2]!] || 0));
  }

  const parsed = new Date(value || "");
  if (Number.isNaN(parsed.getTime())) {
    throw new Error(
      `Invalid date "${value}". Use ISO 8601, "now", or a duration such as 24h or 7d.`,
    );
  }
  return parsed;
}

export interface SubscriptionSummary {
  subscription_status: unknown;
  enterprise: boolean;
  token_usage: number;
  token_limit: number;
  token_remaining: number;
  usage_percent: number;
  rate_limit: number;
  rate_ttl: number;
  auto_renew: boolean;
  ends_at: unknown;
  user: {
    name: unknown;
    email: unknown;
    verified: boolean;
  };
}

export function summarizeSubscriptionAccount(
  account: Record<string, unknown>,
): SubscriptionSummary {
  const tokenLimit = numberOrZero(account.tokenLimit);
  const tokenUsage = numberOrZero(account.tokenUsage);
  const user = recordOrEmpty(account.user);
  return {
    subscription_status: account.stripeStatus ?? null,
    enterprise: Boolean(account.isEnterprise),
    token_usage: tokenUsage,
    token_limit: tokenLimit,
    token_remaining: Math.max(0, tokenLimit - tokenUsage),
    usage_percent:
      tokenLimit > 0 ? Number(((tokenUsage / tokenLimit) * 100).toFixed(2)) : 0,
    rate_limit: numberOrZero(account.rateLimit),
    rate_ttl: numberOrZero(account.rateTtl),
    auto_renew: Boolean(account.isAutoRenew),
    ends_at: account.endsAt ?? null,
    user: {
      name: user.name ?? null,
      email: user.email ?? null,
      verified: Boolean(user.isVerified),
    },
  };
}
